### PR TITLE
【mindspore】update d2l-mindspore for chapter4 and chapter 5

### PR DESCRIPTION
**Pull request by 1114120549**

### CODE REVIEW

**Comment by lvyufeng** [comment]
0.mlp

```python
def relu(x):
    return ops.ReLU()(x)
```

改为函数式

**Comment by 1114120549** [comment]
稍等，我等下把最新的代码push上去，我这边测试了第四章除了第5节 weight decay，其他的都是正常的。



---Original---
From: ***@***.***&gt;
Date: Thu, Jan 19, 2023 09:48 AM
To: ***@***.***&gt;;
Cc: ***@***.******@***.***&gt;;
Subject: Re: [lvyufeng/d2l-mindspore] 【mindspore】update d2l-mindspore for chapter4 and chapter 5 (PR #8)




 
0.mlp
 def relu(x):     return ops.ReLU()(x)
 
改为函数式

File: chapter_04_multilayer-perceptrons/7_kaggle-house-price.ipynb [hunk 1]
```diff
@@ -387,12 +387,30 @@
...
+    "        loss = log_rmse(z, y)\n",
+    "        return loss\n",
+    "\n",
+    "    # 获取梯度函数\n",
+    "    grad_fn = value_and_grad(forward_fn, None, weights=net.trainable_params())\n",
+    "\n",
+    "    # 定义模型单步训练\n",
+    "    def train_one_step(X, Y):\n",
+    "        loss, grads = grad_fn(X, Y)\n",
+    "        loss = ops.depend(loss, optimizer(grads))\n",
```
Comment (lvyufeng) [inline chapter_04_multilayer-perceptrons/7_kaggle-house-price.ipynb:408]: ops.depend不需要

File: chapter_05_deep-learning-computation/0_model-construction.ipynb [hunk 2]
```diff
@@ -215,18 +217,20 @@
...
     "        super().__init__()\n",
-    "        self.rand_weight = Tensor(np.random.rand(20, 20), mindspore.float32)\n",
+    "        # 不计算梯度的随机权重参数。因此其在训练期间保持不变\n",
+    "        self.rand_weight = Tensor(np.random.rand(20, 20), ms.float32)\n",
     "        self.linear = nn.Dense(20, 20)\n",
     "\n",
     "    def construct(self, X):\n",
     "        X = self.linear(X)\n",
+    "        # 使用创建的常量参数以及ReLU和matmul函数\n",
     "        X = ops.ReLU()(ops.matmul(X, self.rand_weight) + 1)\n",
```
Comment (lvyufeng) [inline chapter_05_deep-learning-computation/0_model-construction.ipynb:230]: ReLU用函数式

File: chapter_05_deep-learning-computation/2_custom-layer.ipynb [hunk 3]
```diff
@@ -124,13 +121,11 @@
...
-    "from mindspore import Tensor, Parameter\n",
-    "\n",
     "class MyLinear(nn.Cell):\n",
     "    def __init__(self, in_units, units):\n",
     "        super().__init__()\n",
-    "        self.weight = Parameter(Tensor(np.random.normal(0, 1, (in_units, units)), mindspore.float32))\n",
-    "        self.bias = Parameter(Tensor(np.random.normal(0, 1, (units,)), mindspore.float32))\n",
+    "        self.weight = Parameter(ops.randn((in_units, units)))\n",
+    "        self.bias = Parameter(ops.randn((units,)))\n",
     "\n",
```
Comment (lvyufeng) [inline chapter_05_deep-learning-computation/2_custom-layer.ipynb:129]: ops.ReLU用函数式

File: chapter_05_deep-learning-computation/3_read-write.ipynb [hunk 4]
```diff
@@ -64,7 +62,7 @@
     "        return self.output(ops.ReLU()(self.hidden(x)))\n",
```
Comment (lvyufeng) [inline chapter_05_deep-learning-computation/3_read-write.ipynb:62]: ops.ReLU 函数式

File: chapter_05_deep-learning-computation/4_use-gpu.ipynb [hunk 5]
```diff
@@ -54,6 +54,33 @@
     "!nvidia-smi"
    ]
   },
+  {
```
Comment (lvyufeng) [inline chapter_05_deep-learning-computation/4_use-gpu.ipynb:57]: 这一节先不要了。

File: chapter_04_multilayer-perceptrons/1_mlp-scratch.ipynb [hunk 6]
```diff
@@ -65,30 +62,48 @@
...
+    "\n",
+    "params = [W1, b1, W2, b2]"
    ]
   },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "outputs": [],
+   "source": [
+    "def relu(X):\n",
```
[the remark this answers is not in the document]
Comment (1114120549) [reply chapter_04_multilayer-perceptrons/1_mlp-scratch.ipynb]: 其他框架都是这样写的，他这里相当于手动实现relu，就是用所有数值与0比较，取最大值，和relu函数的功能是一样的，是教程所需。

File: chapter_04_multilayer-perceptrons/6_numerical-stability-and-init.ipynb [hunk 7]
```diff
@@ -1037,18 +1037,17 @@
...
    "source": [
     "%matplotlib inline\n",
-    "import mindspore\n",
-    "import mindspore.numpy as mnp\n",
-    "import mindspore.ops as ops\n",
     "from d2l import mindspore as d2l\n",
+    "import mindspore as ms\n",
+    "from mindspore import ops, Tensor\n",
     "\n",
     "def sigmoid(x):\n",
```
[the remark this answers is not in the document]
Comment (1114120549) [reply chapter_04_multilayer-perceptrons/6_numerical-stability-and-init.ipynb]: 我把tanh,relu,sigmoid全部改成ops.xxx然后放在mindspore.py里面了，这样可以吗？
